frontend/src/query_builder/ExpressionWidget.jsx

```diff
@@ -0,0 +1,465 @@
+import React, { Component, PropTypes } from 'react';
+
+import _ from 'underscore';
+
+import Icon from "metabase/components/Icon.jsx";
+import Popover from "metabase/components/Popover.jsx";
+
+
+const VALID_OPERATORS = new Set(['+', '-', '*', '/']);
+
+const OPERATOR_SUGGESTIONS = [
+    {name: '+'},
+    {name: '-'},
+    {name: '*'},
+    {name: '/'}
+];
+
+const ERROR_MESSAGE_EMPTY_NAME       = 'what should this expression be called?';
+const ERROR_MESSAGE_EMPTY_EXPRESSION = 'enter an expression';
+
+const KEYCODE_ENTER = 13;
+const KEYCODE_UP    = 38;
+const KEYCODE_DOWN  = 40;
+
+
+function tokenizeExpression(expression, i = 0, level = 0) {
```

Comment:
Hmm, I would think that a function like `tokenizeExpression(expression)` would be a reusable function.  To me it makes sense to pull that out into its own library, test it in isolation, and then simply import it and use it in this react component.

Same would go for `getParsedExpression(tokens)` or `tokenAtPosition`.